clusterversion,sql: remove old cluster versions corresponding to 20.2 and 21.1

postamar opened the pull request:
This pull request removes all old cluster versions corresponding to 20.2 and 21.1 which fall under the SQL group's responsibility.

Fixes #66546.

See #66546 and #65200 for an exhaustive (and possibly exhausting) list.

Commit:
clusterversion: remove GeospatialType, Box2DType

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove AlterColumnTypeGeneral

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove UserDefinedSchemas

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove NoOriginFKIndexes

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove MaterializedViews

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove EmptyArraysInInvertedIndexes

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove UniqueWithoutIndexConstraints

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove VirtualComputedColumns

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove NewSchemaChanger

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove SequencesRegclass

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove ImplicitColumnPartitioning

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove ForeignKeyRepresentationMigration

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove ProtectedTsMetaPrivilegesMigration

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove MultiRegionFeatures

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: remove ChangefeedsSupportPrimaryIndexChanges

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

Commit:
clusterversion: rename LongRunningMigrations

Partially addresses #66546 by removing a cluster version and its
associated dependencies which, for any cluster whose version is at
least 21.1, is certain to be active.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67822)
<!-- Reviewable:end -->

Commit:
ccl: remove local-mixed-20.2-21.1 logic tests

The code covered by these tests is exercised in mixed-version clusters
with 20.2 and 21.1 nodes, as the name of the test suite implies. These
are no longer relevant at this stage of the 21.2 release cycle.

Release note: None

postamar commented:
This PR is best reviewed commit-by-commit.

Note: you'll notice a `ccl: remove local-mixed-20.2-21.1 logic tests` commit. This was necessary for removing both `ImplicitColumnPartitioning` and  `MultiRegionFeatures`. I could squash these three commits into one if anyone cares. I'm operating under the assumption that we don't care about mixed 20.2/21.1 tests at this point in `master`.

postamar commented:
The bazel CI failure seems like it can be ignored. It's complaining about a timeout. I ran the bazel test locally and it succeeded after 507 seconds so I'm guessing that the threshold in the CI is a bit low or something.

ajwerner commented on pkg/startupmigrations/migrations.go:
```diff
@@ -281,10 +281,12 @@ var backwardCompatibleMigrations = []migrationDescriptor{
-		// time (4272248e573cbaa4fac436b0ea07195fcd648845). The below is the first
-		// cluster version that was added after the system.tenants table.
-		includedInBootstrap: clusterversion.ByKey(clusterversion.AlterColumnTypeGeneral),
+		// This migration does not have a dedicated cluster version key but was
```
I think you get to just remove the workFn here; this migration is "baked in" so to speak.

postamar replied:
I originally left this in because if I didn't some backup tests would fail, complaining that `system.tenants` doesn't exist. Among other tests, `TestPaginatedBackupTenant`.

 I've looked into this some more and now think I surfaced an unrelated bug. In the bootstrap package, `system.tenants` gets created only if this is the system tenant. However the migration doesn't seem to care about this. Case in point if I change the current diff to add  `if !r.codec.ForSystemTenant() { return nil }` to `createTenantsTable` the test fails.

postamar replied:
The other tests with problems are `TestBackupRestoreInsideTenant/tenant-backup` and `TestBackupRestoreInsideTenant`, somewhat unsurprisingly.

postamar replied:
I came up with what I believe to be an acceptable fix in that in `backup_planning.go` I wrap the query to `system.tenants` in `if p.ExecCfg().Codec.ForSystemTenant() { ... }` which should be a no-op, AFAIK the table is empty for non-system tenants. Like I said it probably shouldn't even exist in those cases, yet it does in these backup tests. Is this an artifact of how we set up those test clusters or a real bug in the startupmigrations? I still haven't been able to tell.

ajwerner commented on pkg/sql/schemachanger/scpb/scpb.proto:
```diff
@@ -97,7 +97,7 @@ message SequenceDependency {
   // where 12345 is the ID of foo.public.seq.
   // Sequences referenced only by its ID have the ability to be renamed.
-  bool by_id = 5 [(gogoproto.customname) = "ByID"];
+  reserved 5; // This field was used until 21.2.
```
We've never actually stored these things, so part of me thinks we can just nix this.

postamar replied:
Will do.

ajwerner commented on pkg/clusterversion/cockroach_versions.go:
```diff
@@ -197,15 +197,17 @@ const (
 	// All versions which have a registered long-running migration must have a
 	// version higher than this version.
-	LongRunningMigrations
+	deletedLongRunningMigrations
```
Why not just delete these?

postamar replied:
Can't remember why, now I don't see why not, so: will do.

ajwerner reviewed (commented):
LGTM with a few comments

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 3 files at r28, 1 of 9 files at r32, 2 of 6 files at r33.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67822#-:-MfeZaZ32MURhLRqJNJj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @postamar)



<!-- Sent from Reviewable.io -->
